src/bst.py

```diff
@@ -0,0 +1,323 @@
+    """post_order(self): will return a generator that will return the values in the tree using post_order traversal, one at a time."""
+    """breadth_first(self): will return a generator that will return the values in the tree using breadth-first traversal, one at a time."""
+    """delete function: removes node, readjusts tree, fixes pointers and pointees."""
+
+    def __init__(self, if_iter=None):
+        """Init of the Binary Search Tree class."""
+        self.root = None
+        self.counter = 0
+        if if_iter:
+            try:
+                for value in if_iter:
+                    self.insert(value)
+            except TypeError:
+                self.insert(if_iter)
+        self._in_order = self._in_order_trav()
+        self._pre_order = self._pre_order_trav()
+        self._post_order = self._post_order_trav()
+        self._breadth_first = self._breadth_first_trav()
+
+    def insert(self, val):
+        """Take a value, inserts into Binary Search Tree at correct placement."""
+        if self.root is None:
+            self.root = Node(val)
+            self.counter += 1
+
+        else:
+            vertex = self.root
+            while True:
+                if val > vertex.value:
+                    if vertex.right:
+                        vertex = vertex.right
+                    else:
+                        vertex.right = Node(val)
+                        self.counter += 1
+                        break
+
+                elif val < vertex.value:
+                    if vertex.left:
+                        vertex = vertex.left
+                    else:
+                        vertex.left = Node(val)
+                        self.counter += 1
+                        break
+                else:
+                    break
+
+    def size(self):
+        """Return size of Binary Search Tree."""
+        return self.counter
+
+    def contains(self, val):
+        """Return True if val is in the BST, False if not."""
+        if self.search(val):
+            return True
+        return False
+
+    def search(self, val):
+        """Return the node containing that value, else None."""
+        vertex = self.root
+        while vertex:
+            if val > vertex.value:
+                if not vertex.right:
+                    return None
+                vertex = vertex.right
+            elif val < vertex.value:
+                if not vertex.left:
+                    return None
+                vertex = vertex.left
+            else:
+                return vertex
+        return None
+
+    def depth(self):
+        """
+        Return an integer representing the total number of levels in the tree.
+
+        If there is one value, the depth should be 1, if two values it will be 2,
+        if three values it may be 2 or three, depending, etc.
+        """
+        return self._calc_depth(self.root)
+
+    def _calc_depth(self, tree):
```

Comment:
We thought it was the entire tree because its recursive and the node contains pointers to every node below it, so in a way its the entire tree as well as a single node